src/shogun/classifier/svm/SVMLight.cpp

```diff
@@ -1584,18 +1584,18 @@ void SVMLight::update_linear_component_mkl(
 	else // hope the kernel is fast ...
 	{
 		float64_t* w_backup = SG_MALLOC(float64_t, num_kernels);
-		float64_t* w1 = SG_MALLOC(float64_t, num_kernels);
+		SGVector<float64_t> weight_vector(num_weights);
 
 		// backup and set to zero
 		for (int32_t i=0; i<num_kernels; i++)
-		{
 			w_backup[i] = old_beta[i] ;
-			w1[i]=0.0 ;
-		}
+		
 		for (int32_t n=0; n<num_kernels; n++)
 		{
-			w1[n]=1.0 ;
-			kernel->set_subkernel_weights(SGVector<float64_t>(w1, num_weights,false));
+			
+			memset(weight_vector,0,num_kernels);
```

Comment:
there are 2 problems here:
 1) you set the whole `weight_vector` to 0 in the beginning of every iteration... hence you are actually nullifying the result of the previous iteration.
 2) there's no need to set the vector 0 in the beginning (above) since if you check SGVector's constructor you'll see that the allocated memory is actually set to 0 right after allocation

Reply:
Ok, so I should just use 
```
loop {
w1[n] = 1.0
/* code */
w1[n] = 0.0 
}
```